<h2 align="center">Feature scaling</h2>

### Problem Statement:  Classify raisins into one of the two categories,
1. Kecimen
1. Besni

In [5]:
import pandas as pd

df = pd.read_excel("Raisin_Dataset.xlsx")
df.sample(5)

,Area,MajorAxisLength,MinorAxisLength,Eccentricity,ConvexArea,Extent,Perimeter,Class
98,48177,305.784503,202.466194,0.749397,49877,0.677643,860.548,Kecimen
789,87252,433.484520,261.066644,0.798307,90160,0.735795,1166.517,Besni
751,182122,620.487220,376.488606,0.794883,187560,0.721830,1695.230,Besni
123,55787,333.703453,226.951208,0.733121,59520,0.688592,977.425,Kecimen
308,67093,353.109872,243.001034,0.725546,68383,0.760692,972.971,Kecimen


In [6]:
df.shape

(900, 8)

There are total 900 records and using all the features that we have available, we will build a classification model by using support vector machine 

### Train Test Split

In [9]:
X = df[["Area", "MajorAxisLength", "MinorAxisLength", "Eccentricity", "ConvexArea", "Extent", "Perimeter"]]
y = df["Class"]

from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=10)

### Scale the Data

In [11]:
from sklearn.preprocessing import StandardScaler, MinMaxScaler

scaler = StandardScaler()
scaler.fit(X)

X_train_scaled = scaler.transform(X_train) ## no need to scale data as it just contain target class name
X_test_scaled = scaler.transform(X_test)

In [12]:
X_train.head()

,Area,MajorAxisLength,MinorAxisLength,Eccentricity,ConvexArea,Extent,Perimeter
722,126149,478.877101,345.919972,0.691521,134778,0.624908,1549.185
500,143386,469.276508,397.310190,0.532160,146328,0.750260,1422.014
403,91764,433.768160,270.286095,0.782133,93480,0.692925,1165.084
163,91831,463.357872,257.951128,0.830714,95560,0.710370,1252.762
579,130868,613.172615,273.181915,0.895271,134139,0.536784,1513.352


In [13]:
X_train_scaled

array([[ 0.98369528,  0.41344215,  1.83005975, ...,  1.06982849,
        -1.39598956,  1.40080884],
       [ 1.42589144,  0.33065747,  2.8586639 , ...,  1.35328748,
         0.94973473,  0.93602325],
       [ 0.10158614,  0.02447351,  0.3162051 , ...,  0.05629692,
        -0.12319515, -0.00300658],
       ...,
       [ 1.94040593,  1.41267887,  1.94025974, ...,  1.87217696,
         0.09689362,  1.66678384],
       [-0.85910087, -0.68228209, -1.1222849 , ..., -0.86451533,
        -0.81612799, -0.8599019 ],
       [-0.22886166, -0.30895625,  0.04957521, ..., -0.25064868,
        -0.39412057, -0.26752331]], shape=(720, 7))

In [14]:
X_test_scaled

array([[-0.0426657 , -0.14506073,  0.24858118, ..., -0.08106412,
         0.35989352, -0.07215554],
       [-0.43652993, -0.58876265, -0.10176729, ..., -0.47199211,
         0.87296583, -0.59258193],
       [-0.85804906, -0.58924598, -1.23173306, ..., -0.8633864 ,
         0.00766612, -0.80434516],
       ...,
       [-0.91228139, -0.85779301, -0.95531349, ..., -0.86908013,
         0.27931669, -0.84557505],
       [ 0.49085832,  0.58207752,  0.4023557 , ...,  0.46462512,
         1.3787713 ,  0.43838459],
       [-0.89270747, -0.84114895, -0.95288942, ..., -0.87688445,
        -0.14619709, -0.83890868]], shape=(180, 7))

### Model Training Using SVM: RBF Kernel: No Scaling

In [16]:
from sklearn.svm import SVC

model = SVC(kernel="rbf")
model.fit(X_train, y_train)

y_pred = model.predict(X_test)

from sklearn.metrics import classification_report

report = classification_report(y_test, y_pred)
print(report)

model.n_iter_

              precision    recall  f1-score   support

       Besni       0.86      0.75      0.80        83
     Kecimen       0.81      0.90      0.85        97

    accuracy                           0.83       180
   macro avg       0.83      0.82      0.82       180
weighted avg       0.83      0.83      0.83       180



array([229], dtype=int32)

### Model Training Using SVM: RBF Kernel: With Scaling

In [18]:
from sklearn.svm import SVC

model = SVC(kernel="rbf")
model.fit(X_train_scaled, y_train)

y_pred = model.predict(X_test_scaled)

from sklearn.metrics import classification_report

report = classification_report(y_test, y_pred)
print(report)

model.n_iter_

              precision    recall  f1-score   support

       Besni       0.91      0.83      0.87        83
     Kecimen       0.87      0.93      0.90        97

    accuracy                           0.88       180
   macro avg       0.89      0.88      0.88       180
weighted avg       0.88      0.88      0.88       180



array([419], dtype=int32)

After scaling the model performance (accuracy, precision, recall) improves. With scaled data training requires a few more iterations but it is not too big, on the other hand precision, recall, accuracy improvements can be very valuable

### Model Training Using SVM: Linear Kernel: No Scaling

In [21]:
from sklearn.svm import SVC

model = SVC(kernel="linear")
model.fit(X_train, y_train)

y_pred = model.predict(X_test)

from sklearn.metrics import classification_report

report = classification_report(y_test, y_pred)
print(report)

model.n_iter_

              precision    recall  f1-score   support

       Besni       0.91      0.88      0.90        83
     Kecimen       0.90      0.93      0.91        97

    accuracy                           0.91       180
   macro avg       0.91      0.90      0.90       180
weighted avg       0.91      0.91      0.91       180



array([85005907], dtype=int32)

### Model Training Using SVM: Linear Kernel: With Scaling

In [23]:
from sklearn.svm import SVC

model = SVC(kernel="linear")
model.fit(X_train_scaled, y_train)

y_pred = model.predict(X_test_scaled)

from sklearn.metrics import classification_report

report = classification_report(y_test, y_pred)
print(report)

model.n_iter_

              precision    recall  f1-score   support

       Besni       0.90      0.84      0.87        83
     Kecimen       0.87      0.92      0.89        97

    accuracy                           0.88       180
   macro avg       0.88      0.88      0.88       180
weighted avg       0.88      0.88      0.88       180



array([2164], dtype=int32)

For linear kernal, scaling the data reduced the number of iterations drastically (down to 1214 from 85005907). The model performance on the other hand is more or less the same. There is some difference but nothing major.